synapse/events/spamcheck.py

```diff
@@ -30,19 +40,116 @@
+        hooks = {
+            hook: getattr(spam_checker, hook)
+            for hook in supported_hooks
+        }
+
+        api.register_spam_checker_callbacks(**hooks)
+
+
 class SpamChecker:
     def __init__(self, hs: "synapse.server.HomeServer"):
-        self.spam_checkers = []  # type: List[Any]
-        api = hs.get_module_api()
-
-        for module, config in hs.config.spam_checkers:
-            # Older spam checkers don't accept the `api` argument, so we
-            # try and detect support.
-            spam_args = inspect.getfullargspec(module)
-            if "api" in spam_args.args:
-                self.spam_checkers.append(module(config=config, api=api))
-            else:
-                self.spam_checkers.append(module(config=config))
+        self._callbacks: Dict[str, List[Callable]] = {}
+
+    def register_callbacks(
+        self,
+        check_event_for_spam: Callable[
+            ["synapse.events.EventBase"], Union[bool, str]
+        ] = None,
+        user_may_invite: Callable[[str, str, str], bool] = None,
+        user_may_create_room: Callable[[str], bool] = None,
+        user_may_create_room_alias: Callable[[str, RoomAlias], bool] = None,
+        user_may_publish_room: Callable[[str, str], bool] = None,
+        check_username_for_spam: Callable[[Dict[str, str]], bool] = None,
+        check_registration_for_spam: Callable[
+            [Optional[dict], Optional[str], Collection[Tuple[str, str]], Optional[str]],
+            bool,
+        ] = None,
+        check_media_file_for_spam: Callable[
+            [ReadableFileWrapper, FileInfo], bool
+        ] = None,
+    ):
+        """Register callbacks from module for each hook."""
+        if check_event_for_spam is not None:
+            self._register_callback("check_event_for_spam", check_event_for_spam)
+
+        if user_may_invite is not None:
+            self._register_callback("user_may_invite", user_may_invite)
+
+        if user_may_create_room is not None:
+            self._register_callback("user_may_create_room", user_may_create_room)
+
+        if user_may_create_room_alias is not None:
+            self._register_callback(
+                "user_may_create_room_alias",
+                user_may_create_room_alias,
+            )
+
+        if user_may_publish_room is not None:
+            self._register_callback("user_may_publish_room", user_may_publish_room)
+
+        if check_username_for_spam is not None:
+            self._register_callback("check_username_for_spam", check_username_for_spam)
+
+        if check_registration_for_spam is not None:
+            self._register_callback(
+                "check_registration_for_spam",
+                check_registration_for_spam,
+            )
+
+        if check_media_file_for_spam is not None:
+            self._register_callback(
+                "check_media_file_for_spam",
+                check_media_file_for_spam,
+            )
```

Comment:
This function feels a bit meh to me because it looks like a big function with a lot of copy/paste but it was the only way I could think of to make it work with type hints on the callbacks signatures (and the method itself is relatively simple so *shrug*)

Reply:
I think it's fine. It's a bit boilerplatey but as you say it's simple enough. Sometimes it's better just to accept the boilerplate and make everything explicit than it is to do anything too clever.